include/picongpu/plugins/openPMD/WriteSpecies.hpp

```diff
@@ -334,33 +398,106 @@ namespace openPMD
                 particlePushName;
             const std::string particlePush( particlePushName() );
-            iteration.setAttribute( "particlePush", particlePush.c_str() );
+            record.setAttribute( "particlePush", particlePush.c_str() );
 
             traits::GetSpeciesFlagName< ThisSpecies, interpolation<> >
                 particleInterpolationName;
             const std::string particleInterpolation(
                 particleInterpolationName() );
-            iteration.setAttribute(
+            record.setAttribute(
                 "particleInterpolation", particleInterpolation.c_str() );
 
             const std::string particleSmoothing( "none" );
-            iteration.setAttribute(
+            record.setAttribute(
                 "particleSmoothing", particleSmoothing.c_str() );
+
+	    // we do this once for a unit dimension output function, do not touch this code
+	    std::vector< float_64 > zeroUnitDimension( 7, 0.0 );
+	    static constexpr ::openPMD::UnitDimension
+	      openPMDUnitDimensions[ 7 ] = {
+	      ::openPMD::UnitDimension::L,
+	      ::openPMD::UnitDimension::M,
+	      ::openPMD::UnitDimension::T,
+	      ::openPMD::UnitDimension::I,
+	      ::openPMD::UnitDimension::theta,
+	      ::openPMD::UnitDimension::N,
+	      ::openPMD::UnitDimension::J };
+	    std::map<::openPMD::UnitDimension, double > zeroUnitMap;
+	    for( unsigned i = 0; i < 7; ++i )
+	      {
+		zeroUnitMap[ openPMDUnitDimensions[ i ] ] = zeroUnitDimension[ i ];
+	      }
+	    // end of do-not-touch code
+
+	    // now we have a map in a writeable format with all zeroes
+	    // for each record copy it and modify the copy, e.g.
+
+	    // const records stuff
+	    ::openPMD::Datatype dataType = ::openPMD::Datatype::DOUBLE;
+	    ::openPMD::Extent extent = {0};
+	    ::openPMD::Dataset dataSet = ::openPMD::Dataset(dataType, extent);
+
+	    // mass
+	    ::picongpu::detail::GetMassOrZero< FrameType > const getMassOrZero;
+	    if( getMassOrZero.hasMassRatio )
+	      {
+		const float_64 mass( getMassOrZero() );
+		auto & massRecord = record["mass"];
+		auto & massComponent = massRecord[::openPMD::RecordComponent::SCALAR];
+		massComponent.resetDataset( dataSet );
+		massComponent.makeConstant( mass );
+	    
+		auto massUnitMap = zeroUnitMap;
+		massUnitMap[ ::openPMD::UnitDimension::M ] = 1.0;
+		massRecord.setUnitDimension( massUnitMap );
+		massComponent.setUnitSI(::picongpu::UNIT_MASS ); 
+		massRecord.setAttribute("macroWeighted", int32_t(false) );
+		massRecord.setAttribute("weightingPower", float_64(1.0) );
+		massRecord.setAttribute("timeOffset", float_64(0.0) );
+	      }
+
+	    // charge
+	    using hasBoundElectrons = typename pmacc::traits::HasIdentifier<
+	      FrameType,
+            boundElectrons
+	      >::type;
+	    ::picongpu::detail::GetChargeOrZero< FrameType > const getChargeOrZero;
+	    if( ! hasBoundElectrons::value && getChargeOrZero.hasChargeRatio )
```

Comment:
Not sure what should be done for species with bound electrons. Could you check how the original HDF5 plugin (from where we took this code) handled it?